mypy/plugins/functools.py

```diff
@@ -0,0 +1,78 @@
+"""Plugin for supporting the functools standard library module."""
+from typing import Dict, Optional
+from typing_extensions import Final
+
+import mypy.plugin
+from mypy.nodes import ARG_POS, ARG_STAR2, Argument, FuncItem, Var
+from mypy.plugins.common import add_method
+from mypy.types import CallableType, Type
+
+
+functools_total_ordering_makers: Final = {
+    'functools.total_ordering',
+}
+
+_ORDERING_METHODS: Final = {
+    '__lt__',
+    '__le__',
+    '__gt__',
+    '__ge__',
+}
+
+
+def functools_total_ordering_maker_callback(ctx: mypy.plugin.ClassDefContext,
+                                            auto_attribs_default: bool = False) -> None:
+    """Add dunder methods to classes decorated with functools.total_ordering."""
+    if ctx.api.options.python_version < (3, 2):
+        ctx.api.fail("functools.total_ordering is not supported in Python 2 or 3.1", ctx.reason)
+        return
+
+    comparison_methods = _analyze_class(ctx)
+    if not comparison_methods:
+        ctx.api.fail('must define at least one ordering operation: < > <= >=', ctx.reason)
+        return
+
+    root = max(comparison_methods)  # prefer __lt__ to __le__ to __gt__ to __ge__
+    root_method = comparison_methods[root]
+    other_type = _find_other_type(root_method)
+    if isinstance(root_method.type, CallableType):
+        ret_type = root_method.type.ret_type
+    else:
+        ret_type = ctx.api.named_type('__builtins__.bool')
+    for additional_op in _ORDERING_METHODS - {root}:
+        if additional_op not in comparison_methods:
+            args = [Argument(Var('other', other_type), other_type, None, ARG_POS)]
+            add_method(ctx, additional_op, args, ret_type)
+
+
+def _find_other_type(method: FuncItem) -> Optional[Type]:
+    """Find the type of the ``other`` argument in a comparision method."""
+    first_arg_pos = 0 if method.is_static else 1
+    cur_pos_arg = 0
+    other_arg = None
+    for arg in method.arguments:
+        if arg.kind == ARG_POS:
+            if cur_pos_arg == first_arg_pos:
+                other_arg = arg
+                break
+
+            cur_pos_arg += 1
+        elif arg.kind != ARG_STAR2:
+            other_arg = arg
+            break
+    return None if other_arg is None else other_arg.type_annotation
+
+
+def _analyze_class(ctx: mypy.plugin.ClassDefContext) -> Dict[str, FuncItem]:
+    """Analyze the class body, its parents, and return the comparison methods found."""
+    # Traverse the MRO and collect ordering methods.
+    comparison_methods = {}  # type: Dict[str, FuncItem]
+    # Skip object because total_ordering does not use methods from object
+    for cls in ctx.cls.info.mro[:-1]:
+        for name in _ORDERING_METHODS:
+            if name in cls.names and name not in comparison_methods:
+                node = cls.names[name].node
+                if isinstance(node, FuncItem):
+                    comparison_methods[name] = node
```

Comment:
What if the method is implemented using a variable (e.g. `__lt__ = lambda ...`)? More generally, if the dunder method exists as any kind of node in the symbol table (not just `FuncItem`), we probably don't want to generate a method. 

Random idea: use `Optional[FuncItem]` as the dictionary value type. If it's `None`, the method is defined, but it's something unexpected. In this case it's okay to fall back to default types for the signature of a generated method.